src/lib/InstantSearch.js

```diff
@@ -41,33 +77,53 @@ const defaultCreateAlgoliaClient = (defaultAlgoliasearch, appId, apiKey) =>
  * @fires Instantsearch#render This event is triggered each time a render is done
  */
 class InstantSearch extends EventEmitter {
-  constructor({
-    appId = null,
-    apiKey = null,
-    indexName = null,
-    numberLocale,
-    searchParameters = {},
-    urlSync = null,
-    routing = null,
-    searchFunction,
-    createAlgoliaClient = defaultCreateAlgoliaClient,
-    stalledSearchDelay = 200,
-  }) {
+  constructor(options) {
     super();
-    if (appId === null || apiKey === null || indexName === null) {
-      const usage = `
-Usage: instantsearch({
-  appId: 'my_application_id',
-  apiKey: 'my_search_api_key',
-  indexName: 'my_index_name'
-});`;
-      throw new Error(usage);
-    }
 
-    const client = createAlgoliaClient(algoliasearch, appId, apiKey);
-    client.addAlgoliaAgent(`instantsearch.js ${version}`);
+    const {
+      appId = null,
+      apiKey = null,
+      indexName = null,
+      numberLocale,
+      searchParameters = {},
+      urlSync = null,
+      routing = null,
+      searchFunction,
+      createAlgoliaClient = defaultCreateAlgoliaClient,
+      stalledSearchDelay = 200,
+      searchClient = null,
+    } = options;
+
+    checkOptions({
+      appId,
+      apiKey,
+      indexName,
+      createAlgoliaClient,
+      searchClient,
+    });
+
+    if (searchClient) {
+      if (typeof searchClient.search !== 'function') {
+        throw new Error(
+          'InstantSearch configuration error: `searchClient` must implement a `search(requests)` method.'
+        );
+      }
+
+      const client = {
+        __proto__: Object.getPrototypeOf(searchClient),
```

Comment:
I don't think that this is standardised properly.

What about doing instead:

```js
if (typeof searchClient.clearCache === 'function') {
  searchClient.clearCache = function() {};
}

if (typeof searchClient.addAlgoliaAgent === 'function') {
  searchClient.addAlgoliaAgent = function() {};
}
```

_or_ doing this check in the places in InstantSearch those two functions are used?

Reply:
I like your second solution, thanks! I needed `__proto__` because `algoliasearch` relies on it. We come back to the [initial solution](https://github.com/algolia/instantsearch.js/pull/2894/commits/b085d776dc41b4a2cc6c5f828a9467aeb89a58eb#diff-bde03140d12ec15ad28983aea70e6f69R102) but with `Object.create()`.

Reply:
> I like your second solution, thanks! I needed __proto__ because algoliasearch relies on it.

Quick question, since this is a custom search client we have here, why do we need to have __proto or Object.create(). Which part of the code will expect this object to be made this way?

Reply:
What if you just tested if methods you need are existing and if that's the case then do not even care about creating a new object (most probably this is already a good client, ours). And only create a new object if those methods are missing

Reply:
Yes, my preferred option would be to check if `addAlgoliaAgent` exists when we are using it, rather than this workaround of making it always callable

Reply:
We could have done this:

```
const client = searchClient;
```

But this kind of assignation is a reference in JavaScript. That means that if ever we do `client.hello = true`, it will mutate `searchClient`. I think we shouldn't risk it. `Object.create()` creates a deep clone of the object.

We could also have done this:

```
const client = { ...searchClient };
```

But this doesn't copy the prototype, on which `algoliasearch` relies (it is not a simple object). We'd lose the reference to `client.search()` (becoming `undefined`).

Therefore, I think that `Object.create()` is the most appropriate thing to do anyway.

Does that answer your question? Do you agree with this decision?

Reply:
> Do you agree with this decision?
Yes and thanks for the explanations, it might be good to add a small comment saying "We don't want to use the same reference so we clone the client no matter if it already have the necessary methods"